File: riveredge-frontend/src/apps/kuaiai/services/knowledge.ts
/**
 * KU-AI 知识库 / 文档 / 切块 API（spec 138，对齐 spec 136 契约）。
 *
 * 产品前缀 `/api/v1/apps/kuaiai/`；请求体禁止夹带 tenantId（后端 extra=forbid）。
 */

import { apiRequest } from '../../../services/api';

const BASE = '/apps/kuaiai';

export interface PagedResult<T> {
  items: T[];
  total: number;
}

export interface PageParams {
  page?: number;
  page_size?: number;
}

/** 知识库行（status 取值 启用|停用） */
export interface KnowledgeBaseOut {
  id: number;
  uuid: string;
  name: string;
  description?: string | null;
  embedding_model_id?: number | null;
  chunk_size?: number | null;
  chunk_overlap?: number | null;
  expand_enabled?: boolean | null;
  status: string;
  created_at: string;
  updated_at: string;
}

export interface KnowledgeBaseCreatePayload {
  name: string;
  description?: string | null;
  embedding_model_id?: number | null;
  chunk_size?: number | null;
  chunk_overlap?: number | null;
  expand_enabled?: boolean | null;
  status?: string;
}

export type KnowledgeBaseUpdatePayload = Partial<KnowledgeBaseCreatePayload>;

export interface KnowledgeBaseOption {
  id: number;
  name: string;
  description?: string | null;
}

export type DocumentStatus = 'pending' | 'parsing' | 'ready' | 'failed';

export interface DocumentListOut {
  id: number;
  uuid: string;
  knowledge_id?: number | null;
  title: string;
  source_type: string;
  file_uuid?: string | null;
  status: DocumentStatus | string;
  chunk_count: number;
  error_message?: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface DocumentOut extends DocumentListOut {
  raw_content?: string | null;
}

export interface DocumentCreatePayload {
  title: string;
  source_type: string;
  file_uuid?: string;
  raw_content?: string;
}

export interface ChunkOut {
  id: number;
  chunk_index: number;
  content: string;
  char_count: number;
  created_at: string;
}

/** 文件来源文档允许的 source_type（按扩展名映射） */
export const DOCUMENT_FILE_SOURCE_TYPES = [
  'pdf',
  'doc',
  'docx',
  'ppt',
  'pptx',
  'xls',
  'xlsx',
  'md',
  'txt',
] as const;

/** 文本直存（raw_content）允许的 source_type */
export const DOCUMENT_RAW_SOURCE_TYPES = ['txt', 'md'] as const;

/** 按文件名扩展名解析 source_type；不在白名单返回 null */
export function sourceTypeFromFileName(fileName: string): string | null {
  const ext = (fileName.split('.').pop() || '').trim().toLowerCase();
  return (DOCUMENT_FILE_SOURCE_TYPES as readonly string[]).includes(ext) ? ext : null;
}

// ==================== 知识库 ====================

export function listKnowledgeBases(params?: PageParams & { keyword?: string }) {
  return apiRequest<PagedResult<KnowledgeBaseOut>>(`${BASE}/knowledge-bases`, { params });
}

export function createKnowledgeBase(data: KnowledgeBaseCreatePayload) {
  return apiRequest<KnowledgeBaseOut>(`${BASE}/knowledge-bases`, { method: 'POST', data });
}

export function getKnowledgeBase(id: number) {
  return apiRequest<KnowledgeBaseOut>(`${BASE}/knowledge-bases/${id}`);
}

export function updateKnowledgeBase(id: number, data: KnowledgeBaseUpdatePayload) {
  return apiRequest<KnowledgeBaseOut>(`${BASE}/knowledge-bases/${id}`, { method: 'PUT', data });
}

export function deleteKnowledgeBase(id: number) {
  return apiRequest<void>(`${BASE}/knowledge-bases/${id}`, { method: 'DELETE' });
}

export function listKnowledgeBaseOptions() {
  return apiRequest<KnowledgeBaseOption[]>(`${BASE}/knowledge-bases/options`);
}

// ==================== 文档 ====================

export function listDocuments(kbId: number, params?: PageParams) {
  return apiRequest<PagedResult<DocumentListOut>>(`${BASE}/knowledge-bases/${kbId}/documents`, {
    params,
  });
}

export function createDocument(kbId: number, data: DocumentCreatePayload) {
  return apiRequest<DocumentOut>(`${BASE}/knowledge-bases/${kbId}/documents`, {
    method: 'POST',
    data,
  });
}

export function getDocument(id: number) {
  return apiRequest<DocumentOut>(`${BASE}/documents/${id}`);
}

export function deleteDocument(id: number) {
  return apiRequest<void>(`${BASE}/documents/${id}`, { method: 'DELETE' });
}

/** 投递异步解析任务；返回 job 状态对象（解析是异步的，调用方直接提示已提交） */
export function parseDocument(id: number) {
  return apiRequest<Record<string, unknown>>(`${BASE}/documents/${id}/parse`, { method: 'POST' });
}

export function listChunks(docId: number, params?: PageParams) {
  return apiRequest<PagedResult<ChunkOut>>(`${BASE}/documents/${docId}/chunks`, { params });
}

File: riveredge-backend/src/apps/kuaiai/schemas/knowledge.py
"""KU-AI 知识库 Pydantic Schemas（KR-D12，S3 Phase 1）。

请求模型 extra=forbid，禁止夹带 tenantId 等未声明字段；响应模型不回
tenant_id / embedding 向量。业务规则（status 闭集、embedding_model_id 同租户
embed 归属、chunk 参数约束、file_uuid/raw_content 二选一）在服务层校验并抛
400，不在 schema 层拦截（pydantic 校验失败是 422）。
"""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class KnowledgeBaseCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str = Field(..., min_length=1, max_length=200, description="知识库名称（组织内唯一）")
    description: Optional[str] = Field(default=None, description="知识库描述")
    embedding_model_id: Optional[int] = Field(
        default=None, ge=1, description="Embedding 模型目录行 ID（须本租户 embed 启用行；空=回落）"
    )
    chunk_size: Optional[int] = Field(
        default=None, ge=1, description="切块大小；空=回落租户可运营配置"
    )
    chunk_overlap: Optional[int] = Field(
        default=None, ge=0, description="切块重叠；约束 >0 且 <chunk_size（服务层）"
    )
    expand_enabled: Optional[bool] = Field(
        default=None, description="检索前有限展开开关；空=回落租户配置（默认开）"
    )
    status: Optional[str] = Field(default=None, max_length=20, description="启用|停用，缺省启用")


class KnowledgeBaseUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: Optional[str] = Field(default=None, min_length=1, max_length=200)
    description: Optional[str] = None
    # 显式传 null 表示清空回落租户配置/默认 embed；省略表示不动
    embedding_model_id: Optional[int] = Field(default=None, ge=1)
    chunk_size: Optional[int] = Field(default=None, ge=1)
    chunk_overlap: Optional[int] = Field(default=None, ge=0)
    expand_enabled: Optional[bool] = None
    status: Optional[str] = Field(default=None, max_length=20)


class KnowledgeBaseOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    uuid: str
    name: str
    description: Optional[str] = None
    embedding_model_id: Optional[int] = None
    chunk_size: Optional[int] = None
    chunk_overlap: Optional[int] = None
    expand_enabled: Optional[bool] = None
    status: str
    created_at: datetime
    updated_at: datetime


class KnowledgeBaseOption(BaseModel):
    """知识库下拉项（档案勾选 knowledge_ids / 对话页选库用）。"""

    id: int
    name: str
    description: Optional[str] = None


class DocumentCreate(BaseModel):
    """文档上传：file_uuid（对象存储读流）与 raw_content（文本直存）二选一，
    服务层校验缺一 400；source_type 决定解析路由（pdf|doc|docx|ppt|pptx|
    xls|xlsx|md|txt）。
    """

    model_config = ConfigDict(extra="forbid")

    title: str = Field(..., min_length=1, max_length=300, description="文档标题")
    source_type: str = Field(..., min_length=1, max_length=20, description="来源类型（解析路由）")
    file_uuid: Optional[str] = Field(
        default=None, max_length=36, description="附件 file_uuid（对象存储读流）"
    )
    raw_content: Optional[str] = Field(
        default=None, max_length=200000, description="原文内容（文本类来源直存）"
    )


class DocumentListOut(BaseModel):
    """文档列表项：DocumentOut 摘去 ``raw_content`` 全文（列表不回大字段）。
    faq_* 遗留列不接产品，不回；不回 tenant_id。"""

    model_config = ConfigDict(from_attributes=True)

    id: int
    uuid: str
    knowledge_id: Optional[int] = None
    title: str
    source_type: str
    file_uuid: Optional[str] = None
    status: str
    chunk_count: int = 0
    error_message: Optional[str] = None
    is_active: bool = True
    created_at: datetime
    updated_at: datetime


class DocumentOut(DocumentListOut):
    """文档详情响应：列表项 + ``raw_content`` 全文（仅 detail 端点回）。"""

    raw_content: Optional[str] = None


class ChunkOut(BaseModel):
    """切块只读浏览：不回 embedding / embedding_vector / tenant_id。"""

    model_config = ConfigDict(from_attributes=True)

    id: int
    chunk_index: int
    content: str
    char_count: int = 0
    created_at: datetime
